### PR TITLE
Chat app reconfiguration

### PR DESCRIPTION
**Pull request by matejpavlovic**
Chat app reconfiguration. This is still a WIP, as it breaks basically everything except the chat app.
It is, however, the first approach to reconfiguration (demonstrated on the chat app).
It will still be improved and the rest of the code will be adapted accordingly.

### CODE REVIEW

File: pkg/deploytest/fakeapp.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ package deploytest
 import (
 	"encoding/binary"
 	"fmt"
+	"github.com/filecoin-project/mir/pkg/pb/commonpb"
```
Comment (dnkolegov) [inline pkg/deploytest/fakeapp.go:12]: don't forget to make the linter happy

File: pkg/iss/checkpoint.go [hunk 2]
```diff
@@ -163,13 +175,15 @@ func (ct *checkpointTracker) applyMessage(msg *isspb.Checkpoint, source t.NodeID
...
+	if ct.stateSnapshotHash == nil {
 		// The message is received too early, put it aside
 		ct.pendingMessages[source] = msg
 		return events.EmptyList()
-	} else if !bytes.Equal(ct.appSnapshotHash, msg.AppSnapshotHash) {
+	} else if !bytes.Equal(ct.stateSnapshotHash, msg.SnapshotHash) {
 		// Snapshot hash mismatch
-		ct.Log(logging.LevelWarn, "Ignoring Checkpoint message. Mismatching app snapshot hash.", "source", source)
+		ct.Log(logging.LevelWarn, "Ignoring Checkpoint message. Mismatching state snapshot hash.", "source", source)
+		fmt.Println(ct.stateSnapshot)
```
Comment (dnkolegov) [inline pkg/iss/checkpoint.go:185]: remove debug output

File: pkg/iss/checkpoint.go [hunk 3]
```diff
@@ -163,13 +175,15 @@ func (ct *checkpointTracker) applyMessage(msg *isspb.Checkpoint, source t.NodeID
...
+		fmt.Println("")
```
Comment (dnkolegov) [inline pkg/iss/checkpoint.go:186]: the same here

File: pkg/iss/config.go [hunk 4]
```diff
@@ -18,9 +18,14 @@ import (
 // To obtain real time delays, these need to be multiplied by the period of the ticker provided to the Node at runtime.
 type Config struct {
 
-	// The IDs of all nodes that execute the protocol.
+	// The IDs of all nodes that execute the protocol in the first epoch.
 	// Must not be empty.
-	Membership []t.NodeID
+	InitialMembership []t.NodeID
```
Comment (dnkolegov) [inline pkg/iss/config.go:23]: Membership in most parts of the codebase is `map[t.NodeID]t.NodeAddress`, maybe we should use another name here?

File: pkg/iss/leaderselectionpolicy.go [hunk 5]
```diff
@@ -21,6 +21,13 @@ type LeaderSelectionPolicy interface {
 
 	// Suspect updates the state of the policy object by announcing it that node `node` has been suspected in epoch `e`.
 	Suspect(e t.EpochNr, node t.NodeID)
+
+	// Reconfigure returns a new LeaderSelectionPolicy based on the state of the current one,
+	// but using a new configuration.
+	// TODO: Use the whole configuration, not just the membership.
+	Reconfigure(membership []t.NodeID) LeaderSelectionPolicy
```
Comment (dnkolegov) [inline pkg/iss/leaderselectionpolicy.go:28]: the same comment about membership for this and all other places.  Maybe nodeIDs? It sounds better with "Reconfigure"

File: samples/chat-demo/app.go [hunk 6]
```diff
@@ -49,89 +82,186 @@ func (chat *ChatApp) ApplyEvents(eventsIn *events.EventList) (*events.EventList,
...
-	// Restore internal state
-	chat.messages = state.Messages
+	// Restore configuration
+	if err := chat.restoreConfiguration(snapshot.Configuration); err != nil {
+		return nil, err
+	}
 
 	// Print new state
-	fmt.Printf("\n CHAT STATE RESTORED. SHOWING ALL CHAT HISTORY FROM THE BEGINNING.\n")
+	fmt.Printf("\nCHAT STATE RESTORED. SHOWING ALL CHAT HISTORY FROM THE BEGINNING.\n\n")
```
Comment (dnkolegov) [inline samples/chat-demo/app.go:233]: debug output, remove or add TODO to remove it

File: samples/chat-demo/app.go [hunk 7]
```diff
@@ -49,89 +82,186 @@ func (chat *ChatApp) ApplyEvents(eventsIn *events.EventList) (*events.EventList,
...
+		for nID, nAddr := range mem.Membership {
+			var err error
+			chat.memberships[e][t.NodeID(nID)], err = multiaddr.NewMultiaddr(nAddr)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	fmt.Printf("Restored app memberships: %d (epoch: %d)\n", len(chat.memberships), chat.currentEpoch)
```
Comment (dnkolegov) [inline samples/chat-demo/app.go:264]: debug output, remove or add TODO to remove it

File: samples/chat-demo/main.go [hunk 8]
```diff
@@ -97,33 +110,45 @@ func run() error {
...
 	// however, will only submit the received requests to its associated Node.
 	reqReceiverAddrs := make(map[t.NodeID]string)
-	for i := range nodeIDs {
-		reqReceiverAddrs[t.NewNodeIDFromInt(i)] = fmt.Sprintf("127.0.0.1:%d", reqReceiverBasePort+i)
+	for nodeID, nodeIP := range addresses {
+		numericID, err := strconv.Atoi(string(nodeID))
+		if err != nil {
+			return fmt.Errorf("node IDs must be numeric in the sample app: %w", err)
+		}
+		reqReceiverAddrs[nodeID] = fmt.Sprintf("%s:%d", nodeIP, reqReceiverBasePort+numericID)
```
Comment (dnkolegov) [inline samples/chat-demo/main.go:151]: use `net.JoinHostPort` instead of fmt.Sprintf

**Review by dnkolegov — APPROVED** [review]
LGTM! Just several "don't forget" comments!

File: samples/chat-demo/app.go [hunk 9]
```diff
@@ -32,13 +37,41 @@ type ChatApp struct {
...
+
+	// The current epoch number.
+	currentEpoch t.EpochNr
+
+	// For each epoch number, stores the corresponding membership.
+	memberships []map[t.NodeID]t.NodeAddress
+
+	// Network transport module used by Mir.
+	// The app needs a reference to it in order to manage connections when the membership changes.
+	transport net.Transport
```
Comment (dnkolegov) [inline samples/chat-demo/app.go:49]: I believe that we have to use a sync primitive here when accessing  `currentEpoch` and `memberships`. I found that in my implementation of the state manager in Eudico.

File: pkg/deploytest/fakeapp.go (hunk 1 above)
Comment (matejpavlovic) [reply pkg/deploytest/fakeapp.go]: Fixed.

File: pkg/iss/checkpoint.go (hunk 2 above)
Comment (matejpavlovic) [reply pkg/iss/checkpoint.go]: Fixed.

File: pkg/iss/checkpoint.go (hunk 3 above)
Comment (matejpavlovic) [reply pkg/iss/checkpoint.go]: Fixed.

File: pkg/iss/config.go (hunk 4 above)
Comment (matejpavlovic) [reply pkg/iss/config.go]: Fixed. (In fact, I needed to use the map of addresses here anyway in the end.)

File: pkg/iss/leaderselectionpolicy.go (hunk 5 above)
Comment (matejpavlovic) [reply pkg/iss/leaderselectionpolicy.go]: Fixed.

File: samples/chat-demo/app.go (hunk 7 above)
Comment (matejpavlovic) [reply samples/chat-demo/app.go]: Fixed.

File: samples/chat-demo/main.go (hunk 8 above)
Comment (matejpavlovic) [reply samples/chat-demo/main.go]: Fixed.

File: samples/chat-demo/app.go (hunk 9 above)
Comment (matejpavlovic) [reply samples/chat-demo/app.go:47]: Now I completely removed `currentEpoch` (the chat app does not need to keep track of it any more) and `memberships` is never accessed concurrently (unless I missed something). So I think we should be fine without synchronizing (at least for the chat app - might be different for the Eudico integration)

**Pull request merged**